File: src/browser/subagent.ts
import { CdpClient, listTargets } from './cdp.js';

/**
 * Instruction sent to the browser sub-agent.
 * Free-form natural-language instruction in, structured outcome out.
 */
export interface BrowserInstruction {
  capability: 'browser_read';
  /** The URL to navigate to and extract data from. */
  url: string;
  /** Free-form instruction describing what data to extract. */
  instruction: string;
  /** Optional CSS selector to scope extraction. */
  selector?: string;
}

/**
 * Structured outcome returned by the browser sub-agent.
 */
export interface BrowserOutcome {
  success: boolean;
  action_taken: string;
  url: string;
  /** Extracted data as structured key-value pairs. */
  extracted?: Record<string, unknown>;
  /** Raw text content extracted from the page or element. */
  textContent?: string;
  /** Page title at time of extraction. */
  pageTitle?: string;
  error?: string;
}

/**
 * Verification result: re-fetch to confirm extracted data.
 */
export interface BrowserVerification {
  verified: boolean;
  actual_url: string;
  actual_title: string;
}

export interface BrowserSubAgent {
  dispatch(instruction: BrowserInstruction): Promise<BrowserOutcome>;
  verify(url: string): Promise<BrowserVerification>;
}

/**
 * Create a browser sub-agent backed by a CDP client.
 * Read-only only: navigates and extracts, never submits forms or clicks buttons.
 *
 * @param cdpPort Chrome debugging port (default 9222)
 */
export function createBrowserSubAgent(cdpPort = 9222): BrowserSubAgent {
  return {
    async dispatch(instruction: BrowserInstruction): Promise<BrowserOutcome> {
      if (instruction.capability !== 'browser_read') {
        return {
          success: false,
          action_taken: instruction.capability,
          url: instruction.url,
          error: `unknown capability: ${instruction.capability}`,
        };
      }

      const cdp = new CdpClient();
      try {
        // Find an available page target or use the debugging endpoint directly
        const targets = await listTargets(cdpPort);
        const pageTarget = targets.find((t) => t.type === 'page');
        if (!pageTarget) {
          return {
            success: false,
            action_taken: 'browser_read',
            url: instruction.url,
            error: 'no browser page target available — is Chrome running with --remote-debugging-port?',
          };
        }

        const wsUrl = `ws://127.0.0.1:${cdpPort}/devtools/page/${pageTarget.id}`;
        await cdp.connect(wsUrl);
        await cdp.navigate(instruction.url);

        const pageTitle = await cdp.getTitle();
        let textContent: string;

        if (instruction.selector) {
          // Extract text from a specific element
          const text = await cdp.evaluate(
            `(() => { const el = document.querySelector(${JSON.stringify(instruction.selector)}); return el ? el.textContent : null; })()`,
          );
          textContent = (text as string) ?? '';
        } else {
          // Extract full page text
          textContent = (await cdp.evaluate('document.body.innerText')) as string ?? '';
        }

        return {
          success: true,
          action_taken: 'browser_read',
          url: instruction.url,
          pageTitle,
          textContent: textContent.slice(0, 5000), // Cap to avoid huge payloads
        };
      } catch (err) {
        return {
          success: false,
          action_taken: 'browser_read',
          url: instruction.url,
          error: (err as Error).message,
        };
      } finally {
        cdp.close();
      }
    },

    async verify(url: string): Promise<BrowserVerification> {
      const cdp = new CdpClient();
      try {
        const targets = await listTargets(cdpPort);
        const pageTarget = targets.find((t) => t.type === 'page');
        if (!pageTarget) {
          return { verified: false, actual_url: '', actual_title: '' };
        }

        const wsUrl = `ws://127.0.0.1:${cdpPort}/devtools/page/${pageTarget.id}`;
        await cdp.connect(wsUrl);
        await cdp.navigate(url);

        const actual_url = await cdp.getUrl();
        const actual_title = await cdp.getTitle();

        return {
          verified: actual_url === url || actual_url.startsWith(url),
          actual_url,
          actual_title,
        };
      } catch {
        return { verified: false, actual_url: '', actual_title: '' };
      } finally {
        cdp.close();
      }
    },
  };
}

/**
 * Fake browser sub-agent for testing — no real browser needed.
 * Accepts canned responses.
 */
export function createFakeBrowserSubAgent(
  responses: Map<string, { title: string; text: string }>,
): BrowserSubAgent {
  return {
    async dispatch(instruction: BrowserInstruction): Promise<BrowserOutcome> {
      if (instruction.capability !== 'browser_read') {
        return {
          success: false,
          action_taken: instruction.capability,
          url: instruction.url,
          error: `unknown capability: ${instruction.capability}`,
        };
      }
      const entry = responses.get(instruction.url);
      if (!entry) {
        return {
          success: false,
          action_taken: 'browser_read',
          url: instruction.url,
          error: `no canned response for URL: ${instruction.url}`,
        };
      }
      return {
        success: true,
        action_taken: 'browser_read',
        url: instruction.url,
        pageTitle: entry.title,
        textContent: entry.text,
      };
    },

    async verify(url: string): Promise<BrowserVerification> {
      const entry = responses.get(url);
      return {
        verified: !!entry,
        actual_url: url,
        actual_title: entry?.title ?? '',
      };
    },
  };
}

File: python/steward/browser/subagent.py
"""Browser sub-agent — read-only + authenticated-read. CDP layer deferred; fakes for tests."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Protocol

READ_CAPABILITY = "browser_read"
AUTHENTICATED_READ_CAPABILITY = "browser_authenticated_read"


class BrowserSubAgent(Protocol):
    async def dispatch(self, instruction: dict[str, Any]) -> dict[str, Any]: ...
    async def verify(self, url: str) -> dict[str, Any]: ...


@dataclass
class FakeBrowserSubAgent:
    """Fake browser sub-agent for testing — no real browser needed.

    Response entry schema:
        {
            "title": str,
            "text": str,
            "reflects_credentials": bool,   # optional — if True, append resolved_creds to text
            "reflects_in_title": bool,      # optional — if True, append resolved_creds to title
        }
    """

    responses: dict[str, dict[str, Any]]

    async def dispatch(self, instruction: dict[str, Any]) -> dict[str, Any]:
        cap = instruction.get("capability")
        if cap == READ_CAPABILITY:
            return self._dispatch_read(instruction)
        if cap == AUTHENTICATED_READ_CAPABILITY:
            return self._dispatch_authenticated_read(instruction)
        return {
            "success": False,
            "action_taken": cap or "unknown",
            "url": instruction.get("url") or instruction.get("target_url", ""),
            "error": f"unknown capability: {cap}",
        }

    def _dispatch_read(self, instruction: dict[str, Any]) -> dict[str, Any]:
        url = instruction.get("url", "")
        entry = self.responses.get(url)
        if not entry:
            return {
                "success": False,
                "action_taken": READ_CAPABILITY,
                "url": url,
                "error": f"no canned response for URL: {url}",
            }
        return {
            "success": True,
            "action_taken": READ_CAPABILITY,
            "url": url,
            "pageTitle": entry["title"],
            "textContent": entry["text"],
        }

    def _dispatch_authenticated_read(self, instruction: dict[str, Any]) -> dict[str, Any]:
        target_url = instruction.get("target_url", "")
        entry = self.responses.get(target_url)
        if not entry:
            return {
                "success": False,
                "action_taken": AUTHENTICATED_READ_CAPABILITY,
                "url": target_url,
                "error": f"no canned response for URL: {target_url}",
            }
        resolved_creds: list[str] = instruction.get("resolved_creds", [])
        text = entry["text"]
        title = entry["title"]
        if entry.get("reflects_credentials"):
            text = text + " [creds reflected: " + " ".join(resolved_creds) + "]"
        if entry.get("reflects_in_title"):
            title = title + " — " + " ".join(resolved_creds)
        return {
            "success": True,
            "action_taken": AUTHENTICATED_READ_CAPABILITY,
            "url": target_url,
            "pageTitle": title,
            "textContent": text,
        }

    async def verify(self, url: str) -> dict[str, Any]:
        entry = self.responses.get(url)
        return {
            "verified": entry is not None,
            "actual_url": url,
            "actual_title": entry["title"] if entry else "",
        }


def create_fake_browser_sub_agent(responses: dict[str, dict[str, Any]]) -> FakeBrowserSubAgent:
    return FakeBrowserSubAgent(responses=responses)
